### PR TITLE
added capability to run execution in sync mode, i.e on current thread

### PR DESCRIPTION
**Pull request by MeirShpilraien**

### CODE REVIEW

**Comment by MeirShpilraien** [comment]
@rafie I still need to add tests for it, but I thought maybe you should start review cause there is lots of changes ..

File: src/execution_plan.h [hunk 1]
```diff
@@ -207,6 +208,7 @@ typedef struct FlatExecutionReader{
     char* reader;
 }FlatExecutionReader;
 
+#define EXECUTION_PULL_SIZE 1
```
Comment (rafie) [inline src/execution_plan.h]: PULL->POOL

File: src/execution_plan.h [hunk 2]
```diff
@@ -216,6 +218,8 @@ typedef struct FlatExecutionPlan{
     FlatExecutionStep* steps;
     void* PD;
     char* PDType;
+    ExecutionPlan* executionPull[EXECUTION_PULL_SIZE];
```
Comment (rafie) [inline src/execution_plan.h]: Pull->Pool

File: src/keys_reader.c [hunk 3]
```diff
@@ -349,13 +378,20 @@ static Record* KeysReader_Next(ExecutionCtx* ectx, void* ctx){
...
+
 static int KeysReader_OnKeyTouched(RedisModuleCtx *ctx, int type, const char *event, RedisModuleString *key){
     Gears_listIter *iter = Gears_listGetIterator(keysReaderRegistration, AL_START_HEAD);
     Gears_listNode* node = NULL;
     while((node = Gears_listNext(iter))){
-        FlatExecutionPlan* fep = Gears_listNodeValue(node);
+        KeysReadeRegisterData* rData = Gears_listNodeValue(node);
         char* keyStr = RG_STRDUP(RedisModule_StringPtrLen(key, NULL));
-        if(!RedisGears_Run(fep, keyStr, NULL, NULL)){
+        if(!RedisGears_Run(rData->fep, rData->mode, keyStr, KeysReader_ExecutionDone, (void*)(rData->mode == ExecutionModeSync ? NULL + 1 : NULL))){
```
Comment (rafie) [inline src/keys_reader.c]: Maybe define a macro like:
#define PRIVATE_DATA(n) (NULL + (n))

File: src/streams_reader.c [hunk 4]
```diff
@@ -5,39 +5,270 @@
...
+    ssrctx->srtctx = srtctx;
+    ssrctx->keyName = RG_STRDUP(keyName);
+    ssrctx->timerIsSet = false;
+    StreamReader_ReadLastId(ctx, ssrctx);
+    Gears_dictAdd(srtctx->singleStreamData, (char*)keyName, ssrctx);
+    return ssrctx;
+}
+
+static void StreamReaderTriggerCtx_Free(StreamReaderTriggerCtx* srtctx){
+    if((--srtctx->refCount) == 0){
```
Comment (rafie) [inline src/streams_reader.c]: Check for !srtctx->refCount?

File: src/streams_reader.c [hunk 5]
```diff
@@ -5,39 +5,270 @@
...
+                                                             const char* keyName,
+                                                             StreamReaderTriggerCtx* srtctx){
+    SingleStreamReaderCtx* ssrctx = RG_ALLOC(sizeof(*ssrctx));
+    ssrctx->numTriggered = 0;
+    ssrctx->srtctx = srtctx;
+    ssrctx->keyName = RG_STRDUP(keyName);
+    ssrctx->timerIsSet = false;
+    StreamReader_ReadLastId(ctx, ssrctx);
+    Gears_dictAdd(srtctx->singleStreamData, (char*)keyName, ssrctx);
+    return ssrctx;
```
Comment (rafie) [inline src/streams_reader.c]: srtctx and ssrctx look too similar, It's a little confusing.

File: src/streams_reader.c (hunk 4 above)
Comment (MeirShpilraien) [reply src/streams_reader.c]: I think its better this way cause its actual number and not a true/false value..

File: src/execution_plan.c [hunk 6]
```diff
@@ -324,35 +317,68 @@ static void FlatExecutionPlan_SerializeReader(FlatExecutionReader* rfep, Gears_B
...
 }
 
-static void FlatExecutionPlan_Serialize(FlatExecutionPlan* fep, Gears_BufferWriter* bw){
-    FlatExecutionPlan_SerializeReader(fep->reader, bw);
-    RedisGears_BWWriteLong(bw, array_len(fep->steps));
+const char* FlatExecutionPlan_Serialize(FlatExecutionPlan* fep, size_t *len){
+    if(fep->serializedFep){
+        // notice that this is not only an optimization,
+        // when calling register on fep we call this function to serialize
+        // The execution might changed durring run (for example, args might change).
```
Comment (rafie) [inline src/execution_plan.c]: during

File: src/execution_plan.c [hunk 7]
```diff
@@ -1219,22 +1270,99 @@ static void FlatExecutionPlan_RegisterKeySpaceEvent(RedisModuleCtx *ctx, const c
...
+    char* data = RedisGears_BRReadBuffer(&br, &dataLen);
+    FlatExecutionPlan* fep = FlatExecutionPlan_Deserialize(data, dataLen);
+
+    RedisGears_ReaderCallbacks* callbacks = ReadersMgmt_Get(fep->reader->reader);
+    assert(callbacks);
+    assert(callbacks->deserializeTriggerArgs);
+
+    void* args = callbacks->deserializeTriggerArgs(&br);
+    ExecutionMode mode = RedisGears_BRReadLong(&br);
     if(!fep){
```
Comment (rafie) [inline src/execution_plan.c]: this is post some fep usage.

Comment (MeirShpilraien) [reply src/execution_plan.c]: Not sure I follow ..

File: src/module.c [hunk 8]
```diff
@@ -302,6 +306,80 @@ static void* RG_GetFlatExecutionPrivateData(ExecutionCtx* ectx){
...
+            RedisGears_ReaderCallbacks* callbacks = holder->callback;
+            if(!callbacks->clear){
+                continue;
+            }
+            callbacks->clear();
+        }
+        Gears_dictReleaseIterator(iter);
+    } else {
+        // when loading keys phase finished, we load the registrations.
+        for(char* readerName = RedisModule_LoadStringBuffer(rdb, NULL) ;
```
Comment (rafie) [inline src/module.c]: RedisModule_LoadStringBuffer() can return NULL.

File: src/streams_reader.c [hunk 9]
```diff
@@ -222,8 +676,131 @@ static Reader* StreamReader_Create(void* arg){
...
+        StreamReader_SerializeArgs(srctx->args, &bw);
+        RedisModule_SaveStringBuffer(rdb, buff->buff, buff->size);
+
+        RedisModule_SaveUnsigned(rdb, srctx->mode);
+    }
+    RedisModule_SaveUnsigned(rdb, 0); // done
+    Gears_listReleaseIterator(iter);
+}
+
+static void StreamReader_RdbLoad(RedisModuleIO *rdb, int encver){
```
Comment (rafie) [inline src/streams_reader.c]: StreamReader_RdbLoad() cannot fail?

File: src/module.c (hunk 8 above)
Comment (MeirShpilraien) [reply src/module.c]: If it returns NULL it first call moduleRDBLoadError(io); and exit...

File: src/streams_reader.c (hunk 9 above)
Comment (MeirShpilraien) [reply src/streams_reader.c]: Load failure is fatal.

File: src/streams_reader.c [hunk 10]
```diff
@@ -5,39 +5,270 @@
...
+    }
+
+    if(id1.first < id2.first){
+        return -1;
+    }
+
+    return 0;
+}
+
+static StreamId StreamReader_ParseStreamId(const char* streamId){
```
Comment (rafie) [inline src/streams_reader.c]: Maybe it's better to have an invalid streamid and to return it on failure?

Comment (MeirShpilraien) [reply src/streams_reader.c]: I can return NULL on error and assert at outside of the function, sounds good?

File: src/streams_reader.c (hunk 4 above)
Comment (MeirShpilraien) [reply src/streams_reader.c]: Assert(srtctx->refCount != 0)

**Pull request merged**